src/client/dfs/dfs_sys.c

```diff
@@ -0,0 +1,1453 @@
+			name_len);
+		name = new_path + dir_name_len + 1;
+	}
+
+	sys_path->path = path;
+	sys_path->path_len = path_len;
+	sys_path->alloc = new_path;
+	sys_path->dir_name = dir_name;
+	sys_path->dir_name_len = dir_name_len;
+	sys_path->name = name;
+	sys_path->name_len = name_len;
+
+	rc = hash_lookup(dfs_sys, sys_path);
+	if (rc != 0) {
+		sys_path_free(dfs_sys, sys_path);
+		return rc;
+	}
+
+	return rc;
+}
+
+int
+dfs_sys_mount(daos_handle_t poh, daos_handle_t coh, int mflags, int sflags,
+	      dfs_sys_t **_dfs_sys)
+{
+	dfs_sys_t	*dfs_sys;
+	int		rc;
+	uint32_t	hash_feats = D_HASH_FT_EPHEMERAL;
+	bool		no_cache = false;
+	bool		no_lock = false;
+
+	if (_dfs_sys == NULL)
+		return EINVAL;
+
+	if (sflags & DFS_SYS_NO_CACHE) {
+		D_DEBUG(DB_TRACE, "dfs_sys_mount(): DFS_SYS_NO_CACHE.\n");
+		no_cache = true;
+		sflags &= ~DFS_SYS_NO_CACHE;
+	}
+	if (sflags & DFS_SYS_NO_LOCK) {
+		D_DEBUG(DB_TRACE, "dfs_sys_mount(): DFS_SYS_NO_LOCK.\n");
+		no_lock = true;
+		sflags &= ~DFS_SYS_NO_LOCK;
+	}
+
+	if (sflags != 0) {
+		D_DEBUG(DB_TRACE, "DFS_SYS mount invalid sflags.\n");
+		return EINVAL;
+	}
+
+	D_ALLOC_PTR(dfs_sys);
+	if (dfs_sys == NULL)
+		return ENOMEM;
+
+	/* Mount dfs */
+	rc = dfs_mount(poh, coh, mflags, &dfs_sys->dfs);
+	if (rc != 0) {
+		D_DEBUG(DB_TRACE, "dfs_mount() failed (%d)\n", rc);
+		D_GOTO(err_dfs_sys, rc);
+	}
+
+	/* Initialize the hash */
+	if (!no_cache) {
+		if (no_lock)
+			hash_feats |= D_HASH_FT_NOLOCK;
+		else
+			hash_feats |= D_HASH_FT_RWLOCK;
+
+		rc = d_hash_table_create(hash_feats, DFS_SYS_HASH_SIZE,
+					 NULL, &hash_hdl_ops,
+					 &dfs_sys->hash);
+		if (rc != 0) {
+			D_DEBUG(DB_TRACE, "d_hash_table_create() failed: "
+				DF_RC"\n", DP_RC(rc));
+			D_GOTO(err_hash, rc = daos_der2errno(rc));
+		}
+	}
+
+	*_dfs_sys = dfs_sys;
+	return rc;
+
+err_hash:
+	dfs_umount(dfs_sys->dfs);
+err_dfs_sys:
+	D_FREE(dfs_sys);
+	return rc;
+}
+
+/**
+ * Unmount a file system with dfs_mount and destroy the hash.
+ */
+int
+dfs_sys_umount(dfs_sys_t *dfs_sys)
+{
+	int		rc;
+	d_list_t	*rlink;
+
+	if (dfs_sys == NULL)
+		return EINVAL;
+
+	if (dfs_sys->hash != NULL) {
+		/* Decrease each reference by one. */
+		while (1) {
+			rlink = d_hash_rec_first(dfs_sys->hash);
+			if (rlink == NULL)
+				break;
+
+			d_hash_rec_decref(dfs_sys->hash, rlink);
+		}
+
+		rc = d_hash_table_destroy(dfs_sys->hash, false);
+		if (rc != 0) {
+			D_DEBUG(DB_TRACE, "d_hash_table_destroy() failed: "
+				DF_RC"\n", DP_RC(rc));
+			D_GOTO(err_hash, rc = daos_der2errno(rc));
+		}
+	}
+
+	rc = dfs_umount(dfs_sys->dfs);
+	if (rc != 0) {
+		D_DEBUG(DB_TRACE, "dfs_umount() failed (%d)\n", rc);
+		D_GOTO(err_umount, rc);
+	}
+
+	D_FREE(dfs_sys);
+	return rc;
+
+err_hash:
+	dfs_umount(dfs_sys->dfs);
+err_umount:
+	D_FREE(dfs_sys);
+	return rc;
+}
+
+int
+dfs_sys_local2global(dfs_sys_t *dfs_sys, d_iov_t *glob)
+{
+	if (dfs_sys == NULL)
+		return EINVAL;
+
+	/* TODO serialize the dfs_sys flags as well */
+	return dfs_local2global(dfs_sys->dfs, glob);
```

Comment:
This would require more logic. Are we okay doing this in a future patch?

Reply:
yea that's fine